node/src/components/chainspec_loader.rs

```diff
@@ -249,32 +288,119 @@ impl ChainspecLoader {
             root_dir,
             reactor_exit,
             next_upgrade,
+            maybe_immediate_switch_block_data: None,
         };
 
         (chainspec_loader, effects)
     }
 
     fn handle_initialize<REv>(
         &mut self,
-        _effect_builder: EffectBuilder<REv>,
-        maybe_highest_block: Option<Box<BlockHeader>>,
+        effect_builder: EffectBuilder<REv>,
+        maybe_highest_block_header: Option<Box<BlockHeader>>,
     ) -> Effects<Event>
     where
-        REv: Send,
+        REv: From<Event> + From<ContractRuntimeRequest> + Send,
     {
-        self.reactor_exit = Some(
-            Self::should_exit_for_upgrade(
-                maybe_highest_block,
-                self.next_upgrade_activation_point(),
-            )
-            .then(|| ReactorExit::ProcessShouldExit(ExitCode::Success))
-            .unwrap_or(ReactorExit::ProcessShouldContinue),
+        // Check if we're not running a version that's already outdated - if it is, we should exit
+        // and upgrade.
+        if Self::should_exit_for_upgrade(
+            maybe_highest_block_header.as_deref(),
+            self.next_upgrade_activation_point(),
+        ) {
+            self.reactor_exit = Some(ReactorExit::ProcessShouldExit(ExitCode::Success));
+            return Effects::new();
+        }
+
+        match maybe_highest_block_header {
+            Some(header)
+                if self
+                    .chainspec
+                    .protocol_config
+                    .is_last_block_before_activation(&header) =>
+            {
+                // This is a valid run immediately after upgrading the node version, we'll need to
+                // create an immediate switch block.
+                trace!("valid run immediately after upgrade");
```

Comment:
I know this is the level we used to use, but I think from experience it would be worthwhile bumping this to `info!`.